dartagnan/src/main/java/com/dat3m/dartagnan/wmm/relation/base/RelRMW.java

```diff
@@ -122,62 +126,67 @@ public TupleSet getMaxTupleSet(){
     }
 
     @Override
-    protected BoolExpr encodeApprox(Context ctx) {
+    protected BooleanFormula encodeApprox(SolverContext ctx) {
+        FormulaManager fmgr = ctx.getFormulaManager();
+		BooleanFormulaManager bmgr = fmgr.getBooleanFormulaManager();
+        IntegerFormulaManager imgr = fmgr.getIntegerFormulaManager();
+        
         // Encode base (not exclusive pairs) RMW
         TupleSet origEncodeTupleSet = encodeTupleSet;
         encodeTupleSet = new TupleSet(Sets.intersection(encodeTupleSet, baseMaxTupleSet));
-        BoolExpr enc = super.encodeApprox(ctx);
+        BooleanFormula enc = super.encodeApprox(ctx);
         encodeTupleSet = origEncodeTupleSet;
 
         // Encode RMW for exclusive pairs
-        BoolExpr unpredictable = ctx.mkFalse();
+		BooleanFormula unpredictable = bmgr.makeFalse();
         for(Thread thread : task.getProgram().getThreads()) {
             for (Event store : thread.getCache().getEvents(storeFilter)) {
-                BoolExpr storeExec = ctx.mkFalse();
+            	BooleanFormula storeExec = bmgr.makeFalse();
                 for (Event load : thread.getCache().getEvents(loadFilter)) {
                     if (load.getCId() < store.getCId()) {
 
                         // Encode if load and store form an exclusive pair
-                        BoolExpr isPair = exclPair(load, store, ctx);
-                        BoolExpr isExecPair = ctx.mkAnd(isPair, store.exec());
-                        enc = ctx.mkAnd(enc, ctx.mkEq(isPair, pairingCond(thread, load, store, ctx)));
+                    	BooleanFormula isPair = exclPair(load, store, ctx);
+                    	BooleanFormula isExecPair = bmgr.and(isPair, store.exec());
+                        enc = bmgr.and(enc, bmgr.equivalence(isPair, pairingCond(thread, load, store, ctx)));
 
                         // If load and store have the same address
-                        BoolExpr sameAddress = ctx.mkEq(((MemEvent)load).getMemAddressExpr(), (((MemEvent)store).getMemAddressExpr()));
-                        unpredictable = ctx.mkOr(unpredictable, ctx.mkAnd(isExecPair, ctx.mkNot(sameAddress)));
+                        BooleanFormula sameAddress = imgr.equal((IntegerFormula)((MemEvent)load).getMemAddressExpr(), (IntegerFormula)(((MemEvent)store).getMemAddressExpr()));
```

Comment:
Might cause problems if we use bitvectors for addresses (the address encoding does check for BitVectors AFAIK)